First test cleanup

DianaVyshotravka opened the pull request.

DianaVyshotravka committed:
defined ruff for code refactoring

DianaVyshotravka committed:
deleted useless code

DianaVyshotravka committed:
modified BaseTest to use pytest predefined methods for setup and teardown and implemented this functionality to test_page_launch.py.
Now all tests in defined directory can be ran using 'pytest gl_autotool/tests' command

DianaVyshotravka committed:
renamed file

veshimed commented on gl_autotool/pages/basic_page.py:
```diff
@@ -1,71 +1,47 @@
-from playwright.sync_api import Page, Locator, ConsoleMessage
```
The file should be called base not basic

veshimed commented on gl_autotool/pages/basic_page.py:
```diff
@@ -1,71 +1,47 @@
+from playwright.sync_api import Page, ConsoleMessage
 
 
 class Base:
```
```suggestion
class BasePage:
```

veshimed commented on gl_autotool/pages/basic_page.py:
```diff
@@ -1,71 +1,47 @@
         """
         Handle console message
-        :param msg: console message object
+
+        Parameters
+        ----------
+        msg
```
The original format was ok. In tis format, you also have to provide description of the param

veshimed commented on gl_autotool/pages/basic_page.py:
```diff
@@ -1,71 +1,47 @@
             print(f"Console Error: {msg.text}")
         else:
             print(f"Console {msg.type}: {msg.text}")
+        return
```
No need for this return here and in other methods that only return None

veshimed commented on gl_autotool/pages/basic_page.py:
```diff
@@ -1,71 +1,47 @@
+        return
 
-    def open_url(self, url: str):
+    def open_url(self, url: str) -> None:
```
This and the methods below that are only one-line wrappers are not needed

DianaVyshotravka replied:
I kept them just to follow the structure with BasePage classe and its inheritors, and I undestand why they are no use. I tried to make some of this wrappers more complex and custom adding log messages or combining two simple wrappers. One was removed. If I simply removed that methods, BasePage class would be practicaly empty, and there would be no need to use its object in tests. With more complex test objectives the methods will get more complex too
Hope now it's better, if any suggestions I'm ready to improve this code

veshimed commented on gl_autotool/tests/base_test.py:
```diff
@@ -3,22 +3,23 @@
-    def set_up(self, browser_type="chromium"):
-        """Preconditions and setup before test execution"""
+    @pytest.fixture(autouse=True)
+    def setup_and_teardown(self, request):
```
Please , document the request param

veshimed commented on gl_autotool/tests/base_test.py:
```diff
@@ -3,22 +3,23 @@
+        """
+        Fixture for setting up and tearing down Playwright resources.
+        """
+        browser_type = getattr(request, "param", "chromium")  # Default to Chromium
```
this should go somewhere at the top of the file into the constant
```
DEFAULT_BROWSER = "chromium"
```

DianaVyshotravka replied:
I also defined browser not as global veriable of  base_test.py, but as attribute of class BaseTests, also for children to reuse if needed

veshimed commented on gl_autotool/tests/test_page_launch.py:
```diff
@@ -0,0 +1,21 @@
+        Test to verify if web page is loaded and gets its title.
+        """
+        base_page = Base(self.page)
+        url = "https://ultimateqa.com/automation"
```
Base URL should be stored in a constant somewhere
```
BASE_URL = "https://ultimateqa.com/automation"
```

DianaVyshotravka replied:
I decided to define base url as attribute of BaseTest class, so that all children (tests that inherit) will be able to also use this variable

veshimed commented on gl_autotool/tests/test_page_launch.py:
```diff
@@ -0,0 +1,21 @@
+        title = base_page.get_title()
+
+        print(f"Page title: {title}")
+        assert title == "Automation Practice - Ultimate QA", "Wrong title"
```
Will that print the actual title in the error message? If no, then update the error message to print it out

veshimed commented on pyproject.toml:
```diff
@@ -22,6 +22,12 @@ testpaths = [
 ]
 
+[tool.ruff]
+target-version = "py312"  # Цільова версія Python
+select = ["E", "F", "W", "C90"]  # Які правила застосовувати
+ignore = ["E501"]  # Які правила ігнорувати (наприклад, довжина рядка)
```
Use English comments :)

DianaVyshotravka committed:
renamed basic to base

DianaVyshotravka committed:
renamed class Base to BasePage

DianaVyshotravka committed:
deleted some comments

DianaVyshotravka committed:
modified the error message

DianaVyshotravka committed:
defined global variable for browser type

DianaVyshotravka committed:
defined global variable for test page url

DianaVyshotravka committed:
instead of global variable some constant values were created as BaseTest class attributes

DianaVyshotravka committed:
some wrapper methods were modified and some - deleted.
if method returns None, than return was removed and docstrings were modified to describe params
structlog were added to create some log messages

DianaVyshotravka committed:
parameter 'request' in BaseTest class method was described in docstring

veshimed commented on gl_autotool/pages/base_page.py:
```diff
@@ -0,0 +1,56 @@
+import structlog
```
```suggestion
from structlog import get_logger
```

veshimed commented on gl_autotool/pages/base_page.py:
```diff
@@ -0,0 +1,56 @@
+import structlog
+from playwright.sync_api import Page, ConsoleMessage
+
+logger = structlog.get_logger("my_logger")
```
It is better to define a logger in the class's constructor, so all the child classes will have it inherited out of the box. Something like that:
```
        self.log = get_logger(self.__class__.__name__)
```

veshimed commented on gl_autotool/pages/base_page.py:
```diff
@@ -0,0 +1,56 @@
+        url : str
+            The URL to navigate to.
+        timeout : int
+            Maximum time to wait for navigation in milliseconds (default: 30s).
```
Better to make them in seconds and convert in the method itself. More user-friendly :)

veshimed commented on gl_autotool/pages/base_page.py:
```diff
@@ -0,0 +1,56 @@
+
+        Parameters
+        ----------
+        url : str
```
No need to specify type in the docstring as it is already documented as type-hint in the method definition (here and everywhere else)

veshimed commented on gl_autotool/tests/test_page_launch.py:
```diff
@@ -11,11 +15,9 @@ def test_page_load(self):
-        base_page.open_url(url)
-        base_page.wait_for_load()
-        title = base_page.get_title()
+        url = self.BASE_URL
```
No need for that. You can use `self.BASE_URL` directly without recreating it

veshimed commented on gl_autotool/tests/base_test.py:
```diff
@@ -3,12 +3,20 @@
 
 
 class BaseTest:
+    BASE_URL = "https://ultimateqa.com/automation"
+    DEFAULT_BROWSER = "chromium"
```
In our project, all class attributes are only lowercase. Upper case we use for either top-level module constants (outside the class) or in Enum classes.

```suggestion
    base_url= "https://ultimateqa.com/automation"
    default_browser = "chromium"
```

veshimed commented on gl_autotool/tests/base_test.py:
```diff
@@ -3,22 +3,31 @@
-        """Preconditions and setup before test execution"""
+        Parameters
+        ----------
+        request : pytest.FixtureRequest
```
```suggestion
        request: pytest.FixtureRequest
```

veshimed commented on gl_autotool/tests/base_test.py:
```diff
@@ -3,22 +3,31 @@
 class BaseTest:
+    BASE_URL = "https://ultimateqa.com/automation"
+    DEFAULT_BROWSER = "chromium"
+    @pytest.fixture(autouse=True)
+    def setup_and_teardown(self, request):
```
Not critical for this particualr MR, but as a best practice, all fixtures can be defined in the `conftest.py` file in the tests folder

DianaVyshotravka replied:
I tried to move fixture into nother folder (conftest.py), it works now, hope I've done it correctly

veshimed replied:
I see now that keeping it in the base class is a lot easier and more convenient, as this fixture sets instance attributes to be used by the test and the pages.

Sorry or trouble but let's move it back to the base class :) All the rest looks good for starters.

DianaVyshotravka replied:
Oh, that's totaly okay. Thanks 
I reverted that commit

DianaVyshotravka committed:
logger was defined inside class BasePage

DianaVyshotravka committed:
parameter timeout now in seconds and got converted inside the method

DianaVyshotravka committed:
data type got removed from parameters description in docstrings

DianaVyshotravka committed:
use URL directly from parent class

DianaVyshotravka committed:
variables to lowercase

DianaVyshotravka committed:
remove spaces before colons in docstrings

DianaVyshotravka committed:
moved fixture to conftest.py

DianaVyshotravka committed:
Revert "moved fixture to conftest.py"

This reverts commit 38fa6bf433bc52af454a85f61722497d608d369f.

veshimed reviewed (approved):
Approving the changes. Will tune it out a bit more in the next PRs
